pkg/lib/cockpit-components-terminal.jsx

```diff
@@ -103,13 +103,89 @@ const themes = {
  *
  * Also it is possible to set up theme by property 'theme'.
  */
+
+export class TerminalState {
+    terminal;
+    element;
+    channel;
+
+    #default_channel_creator;
+
+    constructor(default_channel_creator) {
```

Comment:
The `default_channel_creator` stuff requires a bit more explanation.  Is that a leftover from previous experimentation?  As far as I can see this is never used and you call .connectChannel() on first component mount...

Reply:
It's used by https://github.com/cockpit-project/cockpit-machines/pull/2093. A playground example here should help. I'll make one. And add docs.

When using an external TerminalState object, that object is created ahead of time for all possible terminals in a VM page. We don't want to open the channel at that time, but only on-demand when a Terminal component with that TerminalState actually appears in the page.

When not using an external TerminalState, a concrete channel is passed into the Terminal component.

What about calling it `channel_creator`, dropping the channel argument from `connectChannel`, and initializing the internal state like this:

```
        this.terminal_state = this.props.state || new TerminalState(() => this.props.channel);
```